spec/controllers/products_controller_spec.rb

```diff
@@ -146,4 +143,87 @@ def mock_post_request(name, price, image)
       end
     end
   end
+
+  describe 'GET #edit' do
+    context 'unauthenticated user' do
+      it 'redirects to sign up form' do
+        get :edit, params: { id: 5 }
+        expect(response).to redirect_to(new_user_session_path)
+      end
+    end
+
+    context 'authenticated user' do
+      let(:user) { create(:user) }
+      before { mock_authentication }
+
+      context 'product exists and belongs to user' do
+        let(:product) { create(:product, user: user) }
+        before { get :edit, params: { id: product.id } }
+
+        it 'finds product' do
+          expect(assigns(:product)).to be_a(Product)
+        end
+
+        it 'returns correct "edit" view' do
+          expect(response).to render_template('products/edit')
+        end
+      end
+
+      context 'product doesnt exists in users scope' do
+        it 'returns 404 error' do
+          expect { get(:edit, params: { id: 5 }) }.to raise_error(ActiveRecord::RecordNotFound)
+        end
+      end
+    end
+  end
+
+  describe 'PATCH #update' do
+    context 'unauthenticated user' do
+      it 'redirects to sign up form' do
+        patch :update, params: { id: 5 }
+        expect(response).to redirect_to(new_user_session_path)
+      end
+    end
+
+    context 'authenticated user' do
+      let(:user) { create(:user) }
+      before { mock_authentication }
+
+      context 'product exists and belongs to user' do
+        let(:product) { create(:product, user: user) }
+        before do
+          patch :update, params: {
+            id: product.id,
+            product: {
+              name: 'new_test_name',
+              price: 500,
+              active: false
+            }
+          }
+        end
+
+        it 'finds product' do
+          expect(assigns(:product)).to be_a(Product)
+        end
+
+        it 'changes name correctly' do
+          expect(assigns(:product)).to have_attributes(name: 'new_test_name')
+        end
```

Comment:
Siendo estrictos, estos tests estarían mejor si revisas que el producto efectivamente haya cambiado sus atributos. Para eso habría que hacer esto: `expect(product.reload).to have_attributes(...)`